src/core/telemetry.cpp

```diff
@@ -0,0 +1,260 @@
+    }
+    catch (const asio::system_error &e) {
+        NIXL_DEBUG << "Failed to cancel telemetry write timer: " << e.what();
+        // continue anyway since it's not critical
+    }
+    if (buffer_) {
+        writeEventHelper();
+        buffer_.reset();
+    }
+}
+
+void
+nixlTelemetry::initializeTelemetry() {
+    auto buffer_size = std::getenv(TELEMETRY_BUFFER_SIZE_VAR) ?
+        std::stoul(std::getenv(TELEMETRY_BUFFER_SIZE_VAR)) :
+        DEFAULT_TELEMETRY_BUFFER_SIZE;
+
+    auto folder_path = std::getenv(TELEMETRY_DIR_VAR) ? std::getenv(TELEMETRY_DIR_VAR) : "/tmp";
+
+    auto full_file_path = fs::path(folder_path) / file_;
+
+    if (buffer_size == 0) {
+        throw std::invalid_argument("Telemetry buffer size cannot be 0");
+    }
+
+    NIXL_INFO << "Telemetry enabled, using buffer path: " << full_file_path
+              << " with size: " << buffer_size;
+
+    buffer_ = std::make_unique<sharedRingBuffer<nixlTelemetryEvent>>(
+        full_file_path, true, TELEMETRY_VERSION, buffer_size);
+
+    auto run_interval = std::getenv(TELEMETRY_RUN_INTERVAL_VAR) ?
+        std::chrono::milliseconds(std::stoul(std::getenv(TELEMETRY_RUN_INTERVAL_VAR))) :
+        DEFAULT_TELEMETRY_RUN_INTERVAL;
+
+    // Update write task interval and start it
+    writeTask_.callback_ = [this]() { return writeEventHelper(); };
+    writeTask_.interval_ = run_interval;
+    registerPeriodicTask(writeTask_);
+}
+
+bool
+nixlTelemetry::writeEventHelper() {
+    std::vector<nixlTelemetryEvent> next_queue;
+    // assume next buffer will be the same size as the current one
+    next_queue.reserve(buffer_->capacity());
+    events_.swap(next_queue);
+    for (auto &event : next_queue) {
+        // if full, ignore
+        buffer_->push(event);
+    }
+    // collect all events and sort them by timestamp
+    std::vector<nixlTelemetryEvent> all_events;
+    for (auto &backend : backendMap_) {
+        auto backend_events = backend.second->getTelemetryEvents();
+        for (auto &event : backend_events) {
+            // don't trust enum value coming from backend,
+            // as it might be different from the one in agent
+            event.category_ = nixl_telemetry_category_t::NIXL_TELEMETRY_BACKEND;
+            all_events.push_back(event);
+        }
+    }
+    std::sort(all_events.begin(),
+              all_events.end(),
+              [](const nixlTelemetryEvent &a, const nixlTelemetryEvent &b) {
+                  return a.timestampUs_ < b.timestampUs_;
+              });
+    for (auto &event : all_events) {
+        buffer_->push(event);
+    }
+    return true;
+}
+
+void
+nixlTelemetry::registerPeriodicTask(periodicTask &task) {
+    task.timer_.expires_after(task.interval_);
+    task.timer_.async_wait([this, &task](const asio::error_code &ec) {
+        if (ec != asio::error::operation_aborted) {
+            auto start_time = std::chrono::steady_clock::now();
+
+            if (!task.callback_ || !task.callback_()) {
```

Comment:
What if "callback_ = nullptr" is executed after the first check here but before the call to callback_()? For correctness don't you need a proper check for some atomic cancellation flag before re-arming the timer at line 149?

Reply:
I think task should be added with a bool flag of enabled but in next release.

I think this code has no issues, when class initializes the timer, it runs once and the function is null so we have 2 options regarding the init flow racing the line `writeTask_.callback_ = [this]() { return writeEventHelper(); };` 

1. it happens before\between the if
2. it happens after

if `writeTask_.callback_=` happens before\between, the timer will run with no issues except maybe the first time will be with wrong interval and that will be fixed later when init functions calls registerPeriodicTask. timer is thread safe so the timer will just be armed twice.

if it happens after, the timer expire and will be registered again in when calling registerPeriodicTask(writeTask_);